subprojects/core/src/main/java/org/gradle/internal/service/scopes/DefaultExceptionCollector.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service.scopes;
+
+import groovy.lang.Closure;
+import org.gradle.api.Action;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class DefaultExceptionCollector implements ExceptionCollector {
```

Comment:
I would name this class `KotlinDslExceptionCollector` given it is wired as a service only if the kotlin-dsl specific property is set